FTBFS on kFreeBSD

Zoneminder 1.28.1 FTBFS on kFreeBSD:

~~~~
[ 22%] Building CXX object src/CMakeFiles/zm.dir/zm_logger.cpp.o
cd /«PKGBUILDDIR»/dbuild/src && /usr/bin/c++    -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2  -Wall -D__STDC_CONSTANT_MACROS -O2 -I/«PKGBUILDDIR»/dbuild/src -I/«PKGBUILDDIR»/src -I/«PKGBUILDDIR»/dbuild    -o CMakeFiles/zm.dir/zm_logger.cpp.o -c /«PKGBUILDDIR»/src/zm_logger.cpp
/«PKGBUILDDIR»/src/zm_logger.cpp: In member function 'void Logger::logPrint(bool, const char*, int, int, const char*, ...)':
/«PKGBUILDDIR»/src/zm_logger.cpp:531:29: error: 'SYS_gettid' was not declared in this scope
         if ( (tid = syscall(SYS_gettid)) < 0 ) // Thread/Process id
                             ^
/usr/bin/cmake -E cmake_progress_report /«PKGBUILDDIR»/dbuild/CMakeFiles 12
src/CMakeFiles/zm.dir/build.make:264: recipe for target 'src/CMakeFiles/zm.dir/zm_logger.cpp.o' failed
make[3]: *** [src/CMakeFiles/zm.dir/zm_logger.cpp.o] Error 1
~~~~

See full build logs:

 * https://buildd.debian.org/status/fetch.php?pkg=zoneminder&arch=kfreebsd-amd64&ver=1.28.1-1&stamp=1427780071
 * https://buildd.debian.org/status/fetch.php?pkg=zoneminder&arch=kfreebsd-i386&ver=1.28.1-1&stamp=1427779905

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10256028-ftbfs-on-kfreebsd?utm_campaign=plugin&utm_content=tracker%2F188329&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F188329&utm_medium=issues&utm_source=github).
</bountysource-plugin>

Just tried to build latest `master` with new FreeBSD fixes on kFreeBSD. It still FTBFS:

~~~~
/var/tmp/src/zoneminder/master/src/zm_logger.cpp: In member function ‘void Logger::logPrint(bool, const char*, int, int, const char*, ...)’:
/var/tmp/src/zoneminder/master/src/zm_logger.cpp:541:29: error: ‘SYS_gettid’ was not declared in this scope
         if ( (tid = syscall(SYS_gettid)) < 0 ) // Thread/Process id
                             ^
~~~~

Thanks, @connortechnology. With 209f23c compilation progressed a bit further until it failed as follows:

~~~~
[ 51%] Building CXX object src/CMakeFiles/zm.dir/zm_signal.cpp.o
/var/tmp/src/zoneminder/master/src/zm_signal.cpp: In function ‘void zm_die_handler(int, siginfo_t*, void*)’:
/var/tmp/src/zoneminder/master/src/zm_signal.cpp:75:33: error: ‘struct mcontext_t’ has no member named ‘gregs’
   ip = (void *)(uc->uc_mcontext.gregs[REG_EIP]);
                                 ^
/var/tmp/src/zoneminder/master/src/zm_signal.cpp:75:39: error: ‘REG_EIP’ was not declared in this scope
   ip = (void *)(uc->uc_mcontext.gregs[REG_EIP]);
                                       ^
src/CMakeFiles/zm.dir/build.make:813: recipe for target 'src/CMakeFiles/zm.dir/zm_signal.cpp.o' failed
~~~~

--- a/src/zm_local_camera.cpp
+++ b/src/zm_local_camera.cpp
@@ -32,6 +32,13 @@
 #include <stdlib.h>
 #include <limits.h>
 
+/* Workaround for GNU/kFreeBSD */
+#if defined(__FreeBSD_kernel__)
+#ifndef ENODATA
+#define ENODATA ENOATTR
+#endif
+#endif
+
 static unsigned int BigEndian;
 
 static int vidioctl( int fd, int request, void *arg )

--- a/src/zm_logger.cpp
+++ b/src/zm_logger.cpp
@@ -538,7 +538,12 @@ void Logger::logPrint( bool hex, const char * const file, const int line, const 
         if (tid < 0 ) // Thread/Process id
 #else
 #ifdef HAVE_SYSCALL
+	#ifdef __FreeBSD_kernel__
+        if ( (syscall(SYS_thr_self, &tid)) < 0 ) // Thread/Process id
+
+	# else
         if ( (tid = syscall(SYS_gettid)) < 0 ) // Thread/Process id
+	#endif
 #endif // HAVE_SYSCALL
 #endif
         tid = getpid(); // Process id

--- a/src/zm_monitor.cpp
+++ b/src/zm_monitor.cpp
@@ -515,14 +515,18 @@ bool Monitor::connect() {
         Error( "Got unexpected memory map file size %ld, expected %d", map_stat.st_size, mem_size );
 		return false;
 	} else {
+#ifdef MAP_LOCKED
 		mem_ptr = (unsigned char *)mmap( NULL, mem_size, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_LOCKED, map_fd, 0 );
 		if ( mem_ptr == MAP_FAILED ) {
 			if ( errno == EAGAIN ) {
 				Debug( 1, "Unable to map file %s (%d bytes) to locked memory, trying unlocked", mem_file, mem_size );
+#endif
 				mem_ptr = (unsigned char *)mmap( NULL, mem_size, PROT_READ|PROT_WRITE, MAP_SHARED, map_fd, 0 );
 				Debug( 1, "Mapped file %s (%d bytes) to locked memory, unlocked", mem_file, mem_size );
+#ifdef MAP_LOCKED
 			}
 		}
+#endif
 		if ( mem_ptr == MAP_FAILED )
 			Fatal( "Can't map file %s (%d bytes) to memory: %s(%d)", mem_file, mem_size, strerror(errno), errno );
     }

--- a/src/zm_signal.cpp
+++ b/src/zm_signal.cpp
@@ -72,7 +72,11 @@ RETSIGTYPE zm_die_handler(int signal)
 		ip = (void *)(uc->uc_mcontext.gregs[REG_RIP]);
 	#endif
 #else
+	#ifdef __FreeBSD_kernel__
+		ip = (void *)(uc->uc_mcontext.mc_eip);
+	#else
 		ip = (void *)(uc->uc_mcontext.gregs[REG_EIP]);
+	#endif
 #endif				// defined(__x86_64__)
 
 		// Print the signal address and instruction pointer if available

--- a/src/zm_signal.cpp
+++ b/src/zm_signal.cpp
@@ -64,11 +64,14 @@ RETSIGTYPE zm_die_handler(int signal)
 		      info->si_uid, info->si_status);
 
 		ucontext_t *uc = (ucontext_t *) context;
+		cr2 = info->si_addr;
 #if defined(__x86_64__)
-		cr2 = info->si_addr;
+	#ifdef __FreeBSD_kernel__
+		ip = (void *)(uc->uc_mcontext.mc_rip);
+	#else
 		ip = (void *)(uc->uc_mcontext.gregs[REG_RIP]);
+	#endif
 #else
-		cr2 = info->si_addr;
 		ip = (void *)(uc->uc_mcontext.gregs[REG_EIP]);
 #endif				// defined(__x86_64__)
 

--- a/src/zm_thread.h
+++ b/src/zm_thread.h
@@ -36,15 +36,19 @@ class ThreadException : public Exception
 {
 private:
 pid_t pid() {
-pid_t tid; 
+    pid_t tid; 
 #ifdef __FreeBSD__ 
-long lwpid; 
-thr_self(&lwpid); 
-tid = lwpid; 
+    long lwpid; 
+    thr_self(&lwpid); 
+    tid = lwpid; 
 #else 
-tid=syscall(SYS_gettid); 
+    #ifdef __FreeBSD_kernel__
+        if ( (syscall(SYS_thr_self, &tid)) < 0 ) // Thread/Process id
+    # else
+        tid=syscall(SYS_gettid); 
+    #endif
 #endif
-return tid;
+    return tid;
 }	
 public:
     ThreadException( const std::string &message ) : Exception( stringtf( "(%d) "+message, (long int)pid() ) ) {
@@ -219,13 +223,18 @@ protected:
 
     pid_t id() const
     {
-pid_t tid; 
+        pid_t tid; 
 #ifdef __FreeBSD__ 
-long lwpid; 
-thr_self(&lwpid); 
-tid = lwpid; 
+        long lwpid; 
+        thr_self(&lwpid); 
+        tid = lwpid; 
 #else 
-tid=syscall(SYS_gettid); 
+    #ifdef __FreeBSD_kernel__
+        if ( (syscall(SYS_thr_self, &tid)) < 0 ) // Thread/Process id
+
+    #else
+        tid=syscall(SYS_gettid); 
+    #endif
 #endif
 return tid;
     }

--- a/src/zm_timer.h
+++ b/src/zm_timer.h
@@ -40,7 +40,11 @@ private:
 		thr_self(&lwpid);
 		tid = lwpid;
 #else
+    #ifdef __FreeBSD_kernel__
+                if ( (syscall(SYS_thr_self, &tid)) < 0 ) // Thread/Process id
+    #else
 		tid=syscall(SYS_gettid);
+    #endif
 #endif
 		return tid;
         }

